instance/management/commands/find_orphan_dbs.py

```diff
@@ -0,0 +1,196 @@
+        mysql_parser.add_argument(
+            "-p",
+            "--prompt_password",
+            action='store_true',
+            help="Prompt for password",
+        )
+        mysql_parser.set_defaults(func=self.find_mysql_orphans)
+
+    def handle(self, *args, **options):
+        options['func'](**options)
+    
+    def find_mongo_orphans(
+        self,
+        host,
+        exclude_dbs: list,
+        **kwargs
+    ) -> int:
+        self.stderr.write("Finding orphaned MongoDB databases...")
+
+        mongo_connect_opts = {}
+        if 'user' in kwargs:
+            mongo_connect_opts['username'] = kwargs['user']
+        if kwargs.get('prompt_password'):
+            mongo_connect_opts['password'] = getpass.getpass(
+                stream=self.stderr
+            )
+        mongo = pymongo.MongoClient(host=host, **mongo_connect_opts)
+
+        exclude_dbs = set(exclude_dbs) | MONGO_SYSTEM_DBS
+        LOG.debug("Excluding MongoDB dbs %s", exclude_dbs)
+        non_system_db_names = (
+            db['name'] for db in mongo.list_databases()
+            if db['name'] not in exclude_dbs
+        )
+
+        return self._find_mongo_orphans(host, non_system_db_names)
+    
+    def _find_mongo_orphans(self, host, db_names):
+        count = 0
+        for db_name in db_names:
+            try:
+                MongoDBServer.objects.get(name=db_name, hostname=host)
+            except MongoDBServer.DoesNotExist as ex:
+                self.stderr.write(f"Orphan database '{db_name}' found")
+                count += 1
+                self.stdout.writelines([
+                    f"use {db_name}\n",
+                    "db.dropDatabase()"
+                ])
+            else:
+                LOG.debug("MongoDB DB '%s' is managed by OCIM, skipping", db_name)
+        
+        self.stderr.write(f"{count} orphaned MongoDB databases were found.")
+        return count
+
+    
+    def find_mysql_orphans(
+        self,
+        host,
+        exclude_dbs: list,
+        **kwargs
+    ) -> int:
+        connect_opts={}
+        if kwargs.get('prompt_password'):
+            connect_opts['password'] = getpass.getpass(stream=self.stderr)
+        if 'user' in kwargs:
+            connect_opts['user'] = kwargs['user']
+        exclude_dbs = set(exclude_dbs) | MYSQL_SYSTEM_DBS
+        db = mysql.connect(host=host, **connect_opts)
+        with db.cursor() as cursor:
+            cursor.execute("show databases")
+            count = 0
+            for item in cursor.fetchall():
+                db_name = item[0]
+                if db_name in exclude_dbs:
+                    continue
+                try:
+                    MySQLServer.objects.get(name=db_name, hostname=host)
+                except MySQLServer.DoesNotExist as ex:
+                    self.stderr.write(f"Orphan database '{db_name}' found")
+                    count += 1
+                    self.stdout.write(f"DROP DATABASE {db_name};")
+            return count
```

Comment:
@kewne, this way to determine whether a database is orphan or not is incorrect. `MySQLServer` represents a server, where MySQL is hosted, and we have only four of them. You can even view them in admin panel.

In this context, a correct definition of `database` would be a database schema, which is created when the first app server for some instance is created.

Take a look at [this function](https://github.com/open-craft/opencraft/blob/80596664aa1de6a684a25ae296b135f04d057ba6/instance/models/mixins/database.py#L217-L248). We call it [when spawning an app server](https://github.com/open-craft/opencraft/blob/80596664aa1de6a684a25ae296b135f04d057ba6/instance/models/openedx_instance.py#L304). As a result, we have 11 new database schemas, with names like:
```
| xxx_opencraft_hosting_analytics_api                          |
| xxx_opencraft_hosting_dashboard                              |
| xxx_opencraft_hosting_discovery                              |
| xxx_opencraft_hosting_ecommerce                              |
| xxx_opencraft_hosting_edx_notes_api                          |
| xxx_opencraft_hosting_edxapp                                 |
| xxx_opencraft_hosting_edxapp_csmh                            |
| xxx_opencraft_hosting_notifier                               |
| xxx_opencraft_hosting_programs                               |
| xxx_opencraft_hosting_reports                                |
| xxx_opencraft_hosting_xqueue                                 |
```

I think you're right with the assumption about broken de-provisioning, though.